db/attachment_compaction_test.go

```diff
@@ -802,3 +802,84 @@ func createDocWithInBodyAttachment(t *testing.T, docID string, docBody []byte, a
 
 	return attDocID
 }
+
+// Check for regression of CBG-1980 caused by DCP closing timing issue for the mark and sweep stage
+// May sometimes pass even if there is a regression of the timing issue
+func TestAttachmentCompactIncorrectStat(t *testing.T) {
+	const docsToCreate = 500
+	if base.UnitTestUrlIsWalrus() {
+		t.Skip("Requires CBS")
+	}
+
+	testDb := setupTestDB(t)
+	defer testDb.Close()
+	// Create the docs that will be marked and swept
+	body := map[string]interface{}{"foo": "bar"}
+	for i := 0; i < docsToCreate; i++ {
+		key := fmt.Sprintf("%s_%d", t.Name(), i)
+		_, _, err := testDb.Put(key, body)
+		assert.NoError(t, err)
+	}
+
+	attKeys := make([]string, 0, 11)
+	for i := 0; i < docsToCreate; i++ {
+		docID := fmt.Sprintf("testDoc-%d", i)
+		attKey := fmt.Sprintf("att-%d", i)
+		attBody := map[string]interface{}{"value": strconv.Itoa(i)}
+		attJSONBody, err := base.JSONMarshal(attBody)
+		assert.NoError(t, err)
+		attKeys = append(attKeys, CreateLegacyAttachmentDoc(t, testDb, docID, []byte("{}"), attKey, attJSONBody))
+	}
+
+	base.SetUpTestLogging(t, base.LevelInfo, base.KeyAll)
+	// Start marking stage
+	terminator := base.NewSafeTerminator()
+	stat := &base.AtomicInt{}
+	count := int64(0)
+	var err error
+	go func() {
+		count, _, err = attachmentCompactMarkPhase(testDb, "mark", terminator, stat)
+		require.NoError(t, err)
+	}()
+
+	retryFunc := func() (shouldRetry bool, err error, value interface{}) {
+		if stat.Value() == 0 {
+			return true, nil, nil
+		}
+		return false, nil, nil
+	}
+
+	err, _ = base.RetryLoop("wait for marking to start", retryFunc, base.CreateSleeperFunc(1000, 1))
+	require.NoError(t, err)
+
+	terminator.Close() // Terminate mark function
+	// Allow time for timing issue to be hit where stat increments when it shouldn't
+	time.Sleep(time.Second * 1)
```

Comment:
Keep the sleep here for the test (unfortunately can't think of a better way)
but we should wait for the mark phase to return here.